### PR TITLE
pvr: Series recording

### PR DESCRIPTION
**Pull request by jajoflo**
This is a proposition for series recording  (Replace #533). The idea is to allow the user to record a serie from an EPG and set some simple rules. Non technical user should be able to understand and use the feature.

I did split the PR into 2 commit I just guess that it will be easier like this. If you prefer only one commit i'll sqash them. The skin is very basic i'm sure there plenty a talented skinner could make it look nice.

I did repackage in a new PR to keep thing clean, the most notable change since previous PR is the addition of iClientScheduleId to PVR_TIMER this field will allow the client to store the id of the schedule (i.e Rule) that is the parent of the TIMER.

### CODE REVIEW

**Comment by tsp** [comment]
What about letting the addon's defines the different recording options and use the GUIDialogSettings class to display them? Just a suggestion.

**Comment by jajoflo** [comment]
That was my first idea but then I changed my mind because I think it would be breaking the unified front end concept.  Using capability seam more inline with the rest of the project.

**Comment by tsp** [comment]
I don't think it would break the unified front end concept. It is just a list of different scheduling options and it is a feature that can differ a lot between the different backends. Anyway I will try testing your PR with my mythtv addon and see how it goes.

**Comment by opdenkamp** [comment]
I'll probably have time to review it in the mean time :)

**Comment by adamsutton** [comment]
I've not been keeping up with things on XBMC-PVR, but I'm glad to see someone has started to include series recording :) As I've been updating TVH to include better EPG functionality (inc series recording) and was pestering Lars about this (he must have forgotten about this PR).

I think I agree with the concept of trying to push the more advanced configuration options to the PVR addon's, since they may have all manner of different schemes to aid in series recording. For example I've defined some PVR config options I'd like to include in TVH in the future (Note: this is about more than just series linking): https://docs.google.com/a/adamsutton.me.uk/document/d/1JagS-TDpIyGX2nmVxcUjK9HxpVuBzHvyFTihjdKP_-Y/edit. Some of these are fairly advanced and probably wouldn't be exposed to the average user.

At the very least I think there should be an option to defer decision about HOW series linking is defined to the backend, so the backend server can use its own defined ruleset without exposing to the front end user. Personally this is probably how I would want to use this feature.

Adam

File: xbmc/pvr/addons/PVRClient.cpp [hunk 1]
```diff
@@ -216,12 +218,14 @@ inline void PVRWriteClientTimerInfo(const CPVRTimerInfoTag &xbmcTimer, PVR_TIMER
...
   addonTimer.endTime           = end - g_advancedSettings.m_iPVRTimeCorrection;
   addonTimer.firstDay          = firstDay - g_advancedSettings.m_iPVRTimeCorrection;
-  addonTimer.iEpgUid           = epgTag ? epgTag->UniqueBroadcastID() : -1;
   addonTimer.strSummary        = xbmcTimer.m_strSummary.c_str();
   addonTimer.iMarginStart      = xbmcTimer.m_iMarginStart;
   addonTimer.iMarginEnd        = xbmcTimer.m_iMarginEnd;
   addonTimer.iGenreType        = xbmcTimer.m_iGenreType;
   addonTimer.iGenreSubType     = xbmcTimer.m_iGenreSubType;
+  addonTimer.iSerieRule        = xbmcTimer.m_iSerieRule;
+  addonTimer.iEpgUid           = epgTag ? epgTag->EpgID():0;
```
Comment (opdenkamp) [inline xbmc/pvr/addons/PVRClient.cpp]: this should be using UniqueBroadcastID() instead. EpgID() is the id of the container

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSerie.cpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+#define CONTROL_BTN_THIS_CHANNEL       10
+#define CONTROL_BTN_SAME_WEEKDAY       11
+#define CONTROL_BTN_SAME_TIME          12
+#define CONTROL_BTN_ONCE_A_WEEK        13
+#define CONTROL_BTN_ONCE_A_DAY         14
+#define CONTROL_LBL_WEEK_DAY           15
+
+CGUIDialogPVRTimerSerie::CGUIDialogPVRTimerSerie(void)
+  : CGUIDialog(WINDOW_DIALOG_PVR_TIMER_SERIE, "DialogPVRTimerSerie.xml")
+  , m_progItem(new CFileItem)
```
Comment (opdenkamp) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSerie.cpp]: m_progItem doesn;t seem to be deleted

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSerie.cpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+      thisFlagButton->SetSelected((timer->m_iSerieRule&buttonFlag)!=0);
+    thisFlagButton->SetVisible((dwSupportedRules&buttonFlag)!=0);
+  }
+  return thisFlagButton!=NULL;
+}
+
+void CGUIDialogPVRTimerSerie::OnWindowLoaded()
+{
+  const CEpgInfoTag *tag;
+  tag = m_progItem->GetEPGInfoTag();
```
Comment (opdenkamp) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSerie.cpp]: no need to declare and assign separately. it's not plain C :)

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSerie.cpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+  UpdateButton(CONTROL_BTN_SAME_TIME, PVR_SERIE_SAME_TIME, timer, iSupportedRules);
+  UpdateButton(CONTROL_BTN_ONCE_A_WEEK, PVR_SERIE_ONCE_PER_WEEK, timer, iSupportedRules);
+  UpdateButton(CONTROL_BTN_ONCE_A_DAY, PVR_SERIE_ONCE_PER_DAY, timer, iSupportedRules);
+
+  return CGUIDialog::OnWindowLoaded();
+}
+
+void CGUIDialogPVRTimerSerie::SetProgInfo(const CFileItem *item)
+{
+  *m_progItem = *item;
```
Comment (opdenkamp) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSerie.cpp]: haven't fully reviewed this code yet, but is this perhaps called from another thread?

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSerie.cpp [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+CFileItemPtr CGUIDialogPVRTimerSerie::GetCurrentListItem(int offset)
+{
+  return m_progItem;
+}
+
+bool CGUIDialogPVRTimerSerie::OnClickButtonOK(CGUIMessage &message)
```
Comment (opdenkamp) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSerie.cpp]: you can "const" the message param in all of these calls

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSerie.cpp [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+bool CGUIDialogPVRTimerSerie::OnClickButtonCancel(CGUIMessage &message)
+{
+  bool bReturn = false;
+
+  if (message.GetSenderId() == CONTROL_BTN_CANCEL)
+  {
+    Close();
+    m_iResult=0;
```
Comment (opdenkamp) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSerie.cpp]: just something minor that I've noticed throughout this patch, please put a space between variable, = and value so the coding style matches the one from the other pvr related code.

File: xbmc/pvr/windows/GUIWindowPVRCommon.cpp [hunk 7]
```diff
@@ -702,6 +703,58 @@ bool CGUIWindowPVRCommon::PlayFile(CFileItem *item, bool bPlayMinimized /* = fal
   return true;
 }
 
+bool CGUIWindowPVRCommon::RecordSerie(CFileItem *item)
```
Comment (opdenkamp) [inline xbmc/pvr/windows/GUIWindowPVRCommon.cpp]: const

File: xbmc/pvr/windows/GUIWindowPVRCommon.cpp [hunk 8]
```diff
@@ -702,6 +703,58 @@ bool CGUIWindowPVRCommon::PlayFile(CFileItem *item, bool bPlayMinimized /* = fal
...
+    return false;
+
+  CEpgInfoTag *tag = item->GetEPGInfoTag();
+  if (!tag || !tag->HasPVRChannel())
+    return false;
+
+  CPVRTimerInfoTag *timer = g_PVRTimers->GetMatch(item);
+  if (timer)
+  {
+    CGUIDialogOK::ShowAndGetInput(19033,19034,0,0);
```
Comment (opdenkamp) [inline xbmc/pvr/windows/GUIWindowPVRCommon.cpp]: please put the english translation of the id in a comment next to this. I've been adding that to all new code I'm writing that uses localisation

File: xbmc/pvr/windows/GUIWindowPVRCommon.cpp [hunk 9]
```diff
@@ -702,6 +703,58 @@ bool CGUIWindowPVRCommon::PlayFile(CFileItem *item, bool bPlayMinimized /* = fal
...
+    CGUIDialogYesNo* pDialog = (CGUIDialogYesNo*)g_windowManager.GetWindow(WINDOW_DIALOG_YES_NO);
+    if (!pDialog)
+      return false;
+    pDialog->SetHeading(19257);
+    pDialog->SetLine(0, tag->Title());
+    pDialog->DoModal();
+
+    if (!pDialog->IsConfirmed())
+      return false;
+    newtimer = CPVRTimerInfoTag::CreateFromEpg(*tag);
```
Comment (opdenkamp) [inline xbmc/pvr/windows/GUIWindowPVRCommon.cpp]: needs to be deleted

File: xbmc/pvr/windows/GUIWindowPVRCommon.cpp [hunk 10]
```diff
@@ -702,6 +703,58 @@ bool CGUIWindowPVRCommon::PlayFile(CFileItem *item, bool bPlayMinimized /* = fal
...
+  {
+    CGUIDialogPVRTimerSerie* pDialog = (CGUIDialogPVRTimerSerie*)g_windowManager.GetWindow(WINDOW_DIALOG_PVR_TIMER_SERIE);
+    if (!pDialog)
+      return false;
+    pDialog->SetProgInfo(item);
+    pDialog->DoModal();
+
+    if (!pDialog->IsConfirmed())
+      return false;
+    newtimer = CPVRTimerInfoTag::CreateFromEpg(*tag);
```
Comment (opdenkamp) [inline xbmc/pvr/windows/GUIWindowPVRCommon.cpp]: ditto

**Comment by opdenkamp** [comment]
I agree with @adamsutton and since I'm going to merge in a cleaned up version of the add-on interface anyway, which will force every pvr add-on author to rewrite some things if it wants to be compatible with Frodo+PVR, we should do it properly right away.

I've just done a quick review of the patch, which should be taken care of, and then we can look how we can best expand this to allow some more options.

**Comment by tsp** [comment]
What about adding a record serie button to CGUIDialogPVRGuideInfo? I added it in [my branch](https://github.com/tsp/xbmc/commit/a763e4975d497f521d849d6bc68cdb5d91ceca38 but it is a little quick and dirty fix. 
 Would it make sense to let the PVR addons add the different supported options in the same way as the menuhooks is implemented?

**Comment by fiveisalive** [comment]
In the code all instances of "Serie" should be replaced with "Series".  Assuming that the code is intended to be in English, there is no such word and no distinction between the singular and plural of "Series".  i.e. "I watched the TV series 'Lost'", "I watch  every TV series on ABC": http://en.wikipedia.org/wiki/Television_program#Seasons.2Fseries

**Comment by opdenkamp** [comment]
adding buttons is fine, but let's get the underlying structures and code done first, and then make it look juicy. I just had a quick look at what the patch is doing in that review, not how it's doing it yet. first question about that is why did you choose to use an integer and fool around with that? I'd say you convert that into separate fields, and if you need some int with a mask in the add-on, translate it in there.

and serie should be series indeed :)

**Comment by jajoflo** [comment]
I use a bitflag instead of separate fields because it was  faster to mess around with a bit flag than to type all the fields. But of course before merging it should be convert to separate fields. I'll try to do it this week.

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSerie.cpp (hunk 2 above)
Comment (jajoflo) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSerie.cpp]: m_progItem is a CFileItemPtr...nothing to delete here.

File: xbmc/pvr/windows/GUIWindowPVRCommon.cpp (hunk 8 above)
Comment (jajoflo) [inline xbmc/pvr/windows/GUIWindowPVRCommon.cpp]: Would it be better to def them? Long term it's easier to maintain than in comments.
#define PVR_RES_INFORMATION 19033

File: xbmc/pvr/dialogs/GUIDialogPVRTimerSerie.cpp (hunk 4 above)
Comment (jajoflo) [inline xbmc/pvr/dialogs/GUIDialogPVRTimerSerie.cpp]: It should not but i will have to verify.

**Comment by jajoflo** [comment]
Last commit: 
- Change bitflag to a struct of bool
- Cosmetic
- Sqash
- Rebase

**Comment by adamsutton** [comment]
I still don't like the idea of "guessing" what level of control the backend's can provide for series recording, though I guess some basic options might be ok.

I can't tell from the diff (I'm not familiar enough with XBMC code and haven't had a chance to test it), but are the options at least hidden from the user unless they ask for them? You really don't want to put options in front of the user that may not be supportable by the backends.

Personally I would intend to set up my backend (tvheadend) with its own internal set of series recording rules and just want to have a "record series" button, nothing more. Anything else would just be confusing, but as long as those options are "hidden" by default that would be ok.

**Comment by opdenkamp** [comment]
i agree with @adamsutton about only adding a recordings button at first. but for the long term, we do need some way to configure recording rules in xbmc, so users can do this in 1 gui and don't need to visit a webpage to configure tvheadend (or another tool when using another backend), and this will indeed need a couple of interface changes, e.g. to let xbmc know which capabilities the backend has. but let's do this one step at a time.

**Comment by adamsutton** [comment]
@opdenkamp I think that's what I was trying to get across :)

Just that it may be there are a several ways of handling this:

1) you configure some "default" rules for when you just hit record series (with no other input)
2) option to open advanced settings at the time you hit record series (to manually setup rules for this instance)
3) you configure some profiles to select from (drop down), like "stuff I like to keep" or "news, delete after 7 days", etc...

Each of those 3 above would have access to the same rule sets, they just represent different ways to present this to the user.

I think you'd want at least 1, probably 2 and 3 might be a "nice" to have. For 3 I'd suggest where possible to push the profile storing to the backend.

Also I think the available rules will definitely vary from backend to backend. At the moment the rules available for TVH are quite limited, when compared to say Myth. But that will hopefully change over time.

You don't want to present the user with a set of rules that aren't even supported by the backend. It will be very confusing and generate lots of false issue reports.

**Comment by opdenkamp** [comment]
item 1 should be the focus of this PR indeed.

item 2 and 3 are related, at least in the way that i had in mind. you configure one or more profiles in xbmc, and the gui to configure it should only present options that are available on the backend. item 2 should show the user the same dialog as the configuration dialog for 3, and should create a "temporary profile", that will only be present as long as the timer is present (so the settings can be copied to another timer, if the user wants to do that).

the options that are available will vary from backend to backend indeed, so we need some capabilities field(s) on the interface that describes what the backend does and doesn't support and only display certain options when they are available on the backend, and grey them out otherwise.

however, keep in mind that we support multiple backends at the same time, each with possibly different capabilities, each channel could be served by a different backend, and funky stuff like that. so we need to think this through properly first before writing the code, how we're going to present this in a non-confusing way.

**Comment by adamsutton** [comment]
@opdenkamp I think we're on the same wave length there, at least very close.

As for capabilities in TVH, we will be adding it soon. Especially as we're starting to add several "optional" features. And as we're going through significant dev at the moment I can forsee there will definitely be various TVH versions in the wild offering different support.

Multi-backends and how you cope with them is your problem ;) really you should just drop support for them anyway and stick with the best ;)

But I can see what you mean and how that affects the potential for storing profiles in the backend, so I get what you're saying. Client side profiles will be fine, though maybe when we get down to it we can have both?

But I don't want to set that bar too high ;) I just want the button!

**Comment by opdenkamp** [comment]
multiple backends will be the problem of whoever starts working on these profiles, cause it has to be supported properly or the code won't be included :)

yes, the idea is to have both in the end. i thought of it like this (long term):
- add-on defines which capabilities the backend has, which are different per backend, and could be different per backend version. these capabilities define which options the gui will show for the timer's "advanced settings" and recording profiles
- xbmc will have the capability to define recording profiles, which control things like margins, video transcoding, etc if the backend supports them
- if the backend supports recording profiles too, and the backend provides some way to read/write them remotely, then the add-on will match the profiles in xbmc with the profiles on the backend (optional), so you can sync the profiles.
- the "advanced settings" when creating a new timer will create a temporary recording profile in xbmc only as long as the timer is alive, so it can be copied to another timer, or you can create a permanent profile from it

**Comment by adamsutton** [comment]
All sounds good to me.

I hope to do some HTSP updates very soon. So I'll add capabilities for TVH there, will discuss offline what it might include etc..

**Comment by jajoflo** [comment]
Some clarifications:

1- There's a capabilities field, so the backend can specify wich options it support.

SupportsRecordingRules.bRecord-->true if the back end support series recording.

All the other are optional options: if the option is supported (true) it will be shown otherwise it will be hidden. Currently option are not exclusive but we could eventually add this in the capabilities field. 

SupportsRecordingRules.bAtThisTime; 
SupportsRecordingRules.bOncePerDay; 
SupportsRecordingRules.bOncePerWeek; 
SupportsRecordingRules.bOnThisChannelOnly; 
SupportsRecordingRules.bOnThisWeekDay; 
SupportsRecordingRules.bSkipRepeat; 

@adamsutton: You could set the following has capabilities
SupportsRecordingRules.bRecord = true;
SupportsRecordingRules.bAtThisTime = false; 
SupportsRecordingRules.bOncePerDay = false ; 
SupportsRecordingRules.bOncePerWeek = false; 
SupportsRecordingRules.bOnThisChannelOnly = false; 
SupportsRecordingRules.bOnThisWeekDay = false; 
SupportsRecordingRules.bSkipRepeat = false; 
The dialog would not shown any option.

Multiple backend:
The series recording will use the capabilities and send the command to the backend behind the selected EPG program.

**Comment by adamsutton** [comment]
@jajoflo that helps my understanding, Thanks.

However I still don't want these options visible by default. Many people will want to pre-set these things globally (or leave to backend discretion) and won't want them visible every time they record a series. So something like:

User hits record (anywhere) -> popups Dialog for Single/Series/Advanced -> Single/Series = no more input | Advanced -> Further dialog.

I guess it could be done slightly differently, but most people will want a simple interface to recording series.

From my experience the above flow is how many commercial PVRs work (except the advanced bit, since most are limited to a default set of rules).

Also the decision about "what" features will actually work will sometimes depend from show to show. But that really is getting complicated, so that's up to the backend to figure out.

I'm not sure what the best way to implement this is, since its difficult to second guess what features a backend can support, which will become more interesting if we ever get binary addons.

**Comment by opdenkamp** [comment]
Now that PVR support has been merged into mainline XBMC, every pull request must be sent as a pull request to mainline XBMC too: https://github.com/xbmc/xbmc/pulls

That's why I am closing this one. Sorry for the trouble.

**Pull request closed without merging**